engine/src/flutter/impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -50,496 +52,581 @@ class PositionWriter {
-
-      // For line components, two additional points need to be appended
-      // prior to appending a join connecting the next component.
-      vtx.position = polyline.GetPoint(point_i + 1) + offset_vector;
-      vtx_builder.AppendVertex(vtx.position);
-      vtx.position = polyline.GetPoint(point_i + 1) - offset_vector;
-      vtx_builder.AppendVertex(vtx.position);
-
-      previous_offset = offset;
-      offset = ComputeOffset(point_i + 2, contour_start_point_i,
-                             contour_end_point_i, contour);
-      if (!is_last_component && is_end_of_component) {
-        // Generate join from the current line to the next line.
-        join_proc(vtx_builder, polyline.GetPoint(point_i + 1),
-                  previous_offset.GetVector(), offset.GetVector(),
-                  scaled_miter_limit, scale);
-      }
+ private:
+  PositionWriter& vtx_builder_;
+  const Scalar half_stroke_width_;
+  const Scalar maximum_join_cosine_;
+  const Scalar minimum_miter_cosine_;
+  const Join join_;
+  const Cap cap_;
+  const Scalar scale_;
+  const Tessellator::Trigs& trigs_;
+
+  SeparatedVector2 origin_perpendicular_;
+  Point origin_point_;
+  SeparatedVector2 last_perpendicular_;
+  Point last_point_;
+  bool has_prior_contour_ = false;
+  bool has_prior_segment_ = false;
+  bool contour_needs_cap_ = false;
+
+  // Half of the allowed distance between the ends of the perpendiculars.
+  static constexpr Scalar kJoinPixelThreshold = 0.25f;
+
+  static Scalar ComputeMaximumJoinCosine(Scalar scale,
+                                         Scalar half_stroke_width) {
+    // Consider 2 perpendicular vectors, each pointing to the same side of
+    // two adjacent path segment "boxes". If they are identical, then there
+    // is no turn at that point on the path and we do not need to decorate
+    // that gap with any join geometry. If they differ, there will be a gap
+    // between them that must be decorated and the cosine of the angle of
+    // that gap will be their Dot product (with +1 meaning that there is
+    // no turn and therefore no decoration needed). We need to find the
+    // cosine of the angle between them where we start to care about adding
+    // the join geometry.
+    //
+    // Consider the right triangle where one side is the line bisecting the
+    // two perpendiculars, starting from the common point on the path and
+    // ending at the line that joins them. The hypotenuse of that triangle
+    // is one of the perpendiculars, whose length is (scale * half_width).
+    // The other non-hypotenuse side is kJoinPixelThreshold. This
+    // triangle establishes the equation:
+    //   ||bisector|| ^ 2 + kJoinThreshold ^ 2 == ||hypotenuse|| ^ 2
+    // and the cosine of the angle between the perpendicular and the bisector
+    // will be (||bisector|| / ||hypotenuse||).
+    // The cosine between the perpendiculars which can be compared to the
+    // will be the cosine of double that angle.
```

Comment:
nit: It's worth putting this in the docstring instead.

Reply:
An implementation detail? No.

I do note that this method does not have a doc string, but it would basically read "do magic to determine the maximum cosine that requires a join" and not really explain the magic.

Reply:
Added a public doc comment explaining the requirements of the method, but left the implementation details as an internal comment.